src/analysis/validation.jl

```diff
@@ -2,102 +2,238 @@
 
-            link_dh_wntr[link_name] = Array{Float64,1}(undef,length(solution))
-            link_dh_wm[link_name] = Array{Float64,1}(undef,length(solution))
-        end
+        level_wntr[t] = wntr_simulation.node["pressure"][tank_name].values[t]
+        level_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["tank"][tank_id]["V"]/(1/4*pi*diameter^2)
     end
 
+    tank_df = DataFrame(time = 1:time_step:time_step*duration, volume_wntr = vec(volume_wntr), volume_watermodels = vec(volume_watermodels), 
+        level_wntr = vec(level_wntr), level_watermodels = vec(level_watermodels) )
+
+    return tank_df
+end
+
+# compare links
+function get_link_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, link_id)
+    if link_id in keys(wm_data["nw"]["1"]["pipe"])
+        link_type = "pipe"
+    elseif link_id in keys(wm_data["nw"]["1"]["short_pipe"])
+        link_type = "short_pipe"
+    elseif link_id in keys(wm_data["nw"]["1"]["valve"])
+        link_type = "valve"
+    else
+        link_type = "pump"
+    end
 
-    for tx in 1:length(keys(solution))
-        for ltype in link_types
-            for (key, link) in solution[string(tx)][ltype]
-
-
-                # if ltype in ["check_valve", "shutoff_valve"]
-                #     link_name = data["pipe"][key]["name"]
-                # else
-                #     link_name = data[ltype][key]["name"]
-                # end
-                link_name = data[ltype][key]["name"]
-
-
-                if haskey(wntr_flow, link_name)
-                    link_flow_wntr[link_name][tx] = wntr_flow[link_name].values[tx]
-                    link_flow_wm[link_name][tx] = link["q"]
-
-                    link_dh_wntr[link_name][tx] = abs(wntr_dh[link_name].values[tx])
-                    link_dh_wm[link_name][tx] = link["dhn"]+link["dhp"]
+    duration = length(wm_solution["solution"]["nw"])            # number of time steps
+    time_step = wm_data["time_step"]/3600                       # length per time step (hour)
+    link_name = wm_data["nw"]["1"][link_type][link_id]["source_id"][2]
+    flow_wntr = Array{Float64,1}(undef,duration)
+    flow_watermodels = Array{Float64,1}(undef,duration)
+    head_loss_wntr = Array{Float64,1}(undef,duration)
+    head_loss_watermodels = Array{Float64,1}(undef,duration)
+
+    for t in 1:duration
+        flow_wntr[t] = wntr_simulation.link["flowrate"][link_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)][link_type][link_id]["q"]
+        start_node = wntr_data.links._data[link_name].start_node_name
+        end_node = wntr_data.links._data[link_name].end_node_name
+        head_loss_wntr[t] = wntr_simulation.node["head"][start_node].values[t]-wntr_simulation.node["head"][end_node].values[t]
+        head_loss_watermodels[t] = -1*wm_solution["solution"]["nw"][string(t)][link_type][link_id]["dhn"]+wm_solution["solution"]["nw"][string(t)][link_type][link_id]["dhp"]
```

Comment:
`dhp` and `dhn` are only returned for relaxed problem formulations. This should be replaced with the true difference between heads at adjacent nodes using the `h` solution values.

Reply:
Looks like Yang resolved this.